### PR TITLE
Fixed alpha issues

### LINKED ISSUE #967: draw_sprite_* functions do not obey draw_set_alpha()
They do in GMS, it probably just multiplies the set alpha and the alpha specified by image_alpha or the relevant argument in the sprite. draw_set_alpha() does work correctly with the drawing of normal shapes.

### PR DESCRIPTION
**Pull request by faissaloo**
Fixes #967 and #956 .
Fyi I had to `#include "Graphics_Systems/General/GScolors.h"` so I could use draw_get_alpha() in GSbackground.

### CODE REVIEW

**Comment by RobertBColton** [comment]
```c++
alpha=bind_alpha(alpha)*draw_get_alpha();
```
Should just be:
```
alpha=bind_alpha(alpha);
```

This goes back to another issue, please see my comments on #967 the real fix is this:
```c++
void draw_sprite(int spr, int subimg, gs_scalar x, gs_scalar y, int color = 16777215, gs_scalar alpha = 1);
void draw_sprite(int spr, int subimg, gs_scalar x, gs_scalar y, int color = draw_get_color(), gs_scalar alpha = draw_get_alpha());
```

**Comment by faissaloo** [comment]
No, what @polygoned did with regards to the alpha was correct because alpha has a combining effect in GM, with your fix if you were to specify an alpha it would override the one set by draw_set_alpha().  
As for color, the color set by draw_set_color() only affects ["fonts, forms, primitives and 3D"](http://docs.yoyogames.com/source/dadiospice/002_reference/drawing/color%20and%20blending/draw_set_color.html) and so using it as the default for anything else would cause incompatibility, the default should be c_white for those non-affected functions.

**Comment by RobertBColton** [comment]
I am a little bit confused so I went back and tested again, but couldn't figure this out. Which drawing functions have a combining effect?

For example, when I do the following in GMS, the set alpha does not seem to be multiplied to the passed alpha in the last call and the sprite appears fully opaque:
```
draw_set_alpha(1);
// fully opaque
draw_sprite(sprite0, 0, 0, 0);
draw_set_alpha(0.5);
// half opaque
draw_sprite(sprite0, 0, 0, 30);
// fully opaque
draw_sprite_ext(sprite0, 0, 0, 60, 1, 1, 0, c_white, 1);
draw_set_alpha(1);
// fully opaque
draw_sprite(sprite0, 0, 0, 0);
draw_set_alpha(0.5);
// half opaque
draw_sprite(sprite0, 0, 0, 30);
// fully opaque
draw_sprite_ext(sprite0, 0, 0, 60, 1, 1, 0, c_white, 1);
```

**Comment by faissaloo** [comment]
You're right, turns out I must've imagined it ¯\_(ツ)_/¯ I'll go fix that now.

**Comment by faissaloo** [comment]
Would you also like me to correct the defaults for drawing functions?

**Comment by faissaloo** [comment]
Btw, the draw_sprite* and draw_background* functions don't use draw_set_color(), they use c_white, which I just tested. They do use draw_set_alpha() though.

**Comment by RobertBColton** [comment]
You are right, I went back and tested that code again with draw_set_color(c_yellow); at the top and none of them were affected. So for sprites and backgrounds the default color should be c_white, and the default alpha should be draw_get_alpha. For consistency let's do the same for draw_sprite_ext where we make the arguments optional. Then in each of the functions we can use bind_alpha, without multiplying it, to normalize the alpha between 0 and 1.

I noticed @polygoned never touched fonts, so I think we should just leave them alone. But if we did override the font functions to make color and alpha an option for all of them, I would set the default color to draw_get_color and the alpha to draw_get_alpha.

So to summarize:
* Sprites and backgrounds use c_white as the default color unless the user supplies one.
* Sprites and backgrounds use draw_get_alpha unless the user supplies an alpha.
* Fonts will use draw_get_color and draw_get_alpha if the user supplies only one or neither.

Then our API will be compatible with GMS and legacy GM as well as be consistent. Thoughts? Let me know and I'm going to go test the other drawing functions real fast.

**Comment by faissaloo** [comment]
>You are right, I went back and tested that code again with draw_set_color(c_yellow); at the top and none of them were affected. So for sprites and backgrounds the default color should be c_white, and the default alpha should be draw_get_alpha. For consistency let's do the same for draw_sprite_ext where we make the arguments optional. Then in each of the functions we can use bind_alpha, without multiplying it, to normalize the alpha between 0 and 1.  
  
This is all done.  
  
>I noticed @polygoned never touched fonts, so I think we should just leave them alone. But if we did override the font functions to make color and alpha an option for all of them, I would set the default color to draw_get_color and the alpha to draw_get_alpha.  
  
For now, I think we should keep it how it is (without any defaults) but should, after this merge further talk about it, maybe in an issue or something, because I for one am for it because it would make it consistent with sprites and backgrounds, but I would like to know how everyone else feels about it.  

>Fonts will use draw_get_color and draw_get_alpha if the user supplies only one or neither.  
>Shape functions such as circle/line/rectangle will behave the same as fonts.  
  
It's important to note that neither of these arguments exist in any of the font functions yet and are missing in some of the shape functions.

**Comment by faissaloo** [comment]
Bump.

**Comment by faissaloo** [comment]
Is this substantial enough to warrant adding myself to the header? It says I should ask here:  
https://enigma-dev.org/docs/Wiki/Coding_conventions#Legalese

**Comment by JoshDreamland** [comment]
Up to you; the change spans lots of lines, but it's trivial.

**Comment by faissaloo** [comment]
I just tested the 3D stuff, the behaviour is correct, so no need to change it. It should be ready for merge now unless anyone has anything else to say.

**Comment by RobertBColton** [comment]
Sorry I've been neglecting this from school work. I wanted to pull it locally first and test it, I'll make sure to do that tonight. I just want to double check some drawing edge cases. I assume you've tested various games to make sure they still render properly?

**Comment by faissaloo** [comment]
I suspected you'd be concentrating on school around about now lol, I've tested the basic functions and I've been actively using it as my main copy to work on some games and I've not noticed any issues so far, you'd do best to test it yourself just in case though.

**Comment by RobertBColton** [comment]
This is really weird, I know it's the master branch now. Everything was working fine, I just switched back from faissaloo-master and then pulled from enigma-dev/enigma-dev and now I get the JVM crashes again:

> Owner@BEDROOM-DESKTOP ~/Desktop/ENIGMA/enigma-dev (master)
> $ git remote add upstream https://github.com/enigma-dev/enigma-dev.git
> 
> Owner@BEDROOM-DESKTOP ~/Desktop/ENIGMA/enigma-dev (master)
> $ git pull upstream master
> From https://github.com/enigma-dev/enigma-dev
>  * branch            master     -> FETCH_HEAD
>  * [new branch]      master     -> upstream/master
> Merge made by the 'recursive' strategy.
>  .../SHELL/Graphics_Systems/General/GSstdraw.cpp    | 23 +++++++++-------------
>  1 file changed, 9 insertions(+), 14 deletions(-)
> 
> Owner@BEDROOM-DESKTOP ~/Desktop/ENIGMA/enigma-dev (master)
> $

**Comment by RobertBColton** [comment]
Just ignore me, I deleted ProgramData and compileEGMf and it works again now. Still LGM hanging up with the compile button sometimes, which I know about and need to get fixed.

Anyway, the circle and text drawing examples I posted earlier in this example worked on master. However, the sprite drawing example did not work, but it does work on faisaloo's branch. So just let me run a few more tests.

**Comment by RobertBColton** [comment]
Alright every example seems to be working but I've run onto like a ton more issues along the way.

* MIDI sounds still don't play on master with any sound system but I don't think they ever did.
* I am having a font DPI issues. I know exactly where this code is. My high DPI monitor is making fonts really huge where GMS does not factor DPI into font sizes. You could say LGM is the correct behavior but that makes it impossible to compile your game for other platforms with a lower DPI.
* Somewhere we must have broke something in my inheritance code because the platform example on the EDC no longer works.
http://enigma-dev.org/edc/games.php?game=69
* Elapsed time calculations appear to be off because my God of Man game that used to work perfectly now lags a lot even though it renders at 60fps.
* The GMS cubes example has some problems with its depth buffer in OpenGL1.1 it looks like intense z-fighting. No problems with Direct3D9 though.

These issues are unrelated to this ticket but I want to write them down before I forget them.

My other tests show that this PR fixes both of the original issues. The drawing functions seem to behave much closer to GMS with faissaloo's PR than on the current master. Several games continue to work great including the cubes demo from GMS.

**Comment by RobertBColton** [comment]
I should also mention this is a duplicate of #764 and #956 is a duplicate of #757.

**Comment by RobertBColton** [comment]
@JoshDreamland Everything is looking good, but since that macro is defined all over the place to be the exact same thing, maybe we should put it in the General/GScolors.h file? Like in every graphics systems color.cpp file:
https://github.com/enigma-dev/enigma-dev/blob/51c361c968d4a3c6bd0c934735df453573a290be/ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9colors.cpp#L30

This would work if we moved the macro there because we are already including it in order to use draw_get_alpha.

**Comment by RobertBColton** [comment]
My other tests show that this PR fixes both of the original issues. The drawing functions seem to behave much closer to GMS with faissaloo's PR than on the current master. Several games continue to work great including the cubes demo from GMS.

This PR may also fix an issue in someone's game:
http://enigma-dev.org/forums/index.php?topic=1396

@JoshDreamland @faissaloo Now I should mention this other issue.
http://enigma-dev.org/forums/index.php?topic=1651
While this PR does bring us closer to GMS and GM8.1 behavior, the functions will continue to have differences from GM8.1 and lower. This PR brings it closer to GM8.1 by ensuring that image_alpha works correctly. Where GMS and GM8.1 differ from my own tests is that draw_background(ind, x, y); is not effected by draw_set_alpha in GM8.1 but is effected by it in GMS.

In ENIGMA's master branch and GM8.1 the two functions are drawn fully opaque.
```gml
draw_set_alpha(0.5);
draw_sprite(sprite0,0,256,0);
draw_background(background0,0,0);
```
In GMS and Faissaloo's branch the two functions are drawn half transparent.
```gml
draw_set_alpha(0.5);
draw_sprite(sprite0,0,256,0);
draw_background(background0,0,0);
```

I got GM8.1 working in virtual box and used it to confirm this is the case.
Set Alpha GM8.1

This is why @polygoned requested a setting and I ignored him opting to just keep it the same as GM8.1. So I guess we need to change all those defaults to a macro such as #define DEFAULT_BACKGROUND_ALPHA 1 or #define DEFAULT_BACKGROUND_ALPHA draw_get_alpha()

**Comment by faissaloo** [comment]
@RobertBColton  
>maybe we should put it in the General/GScolors.h file?  
  
It would conflict with the bind_alpha in GLcolors.cpp, I could always just rename the version of bind_alpha() I'm using to bind_alpha_0_to_1() or something though.  
  
>So I guess we need to change all those defaults to a macro such as #define DEFAULT_BACKGROUND_ALPHA 1 or #define DEFAULT_BACKGROUND_ALPHA draw_get_alpha()  
  
Should I put the macros in the .h files themselves or put them in separate .h files like @JoshDreamland is doing [here](https://github.com/enigma-dev/enigma-dev/blob/d9746dfc6d557c4c22c8bd8cf24e02e3272eade0/ENIGMAsystem/SHELL/Collision_Systems/None/Info/collision_info.h)?

**Comment by RobertBColton** [comment]
I'd wait until tomorrow and see what he says because I have no opinion on that. But what we will probably do is add either a checkbox or a text field to the ENIGMA settings frame and I'll show you how to do that.

**Comment by faissaloo** [comment]
bump.

**Comment by JoshDreamland** [comment]
Only recently decided to just drop my GitHub backlog, because it was too large and noisy for any human to process. Appologies.

Anyway, those info files you linked are to aid the build system in figuring out what's going on without parsing the entirety of each system. Virtually nothing belongs in them; I'm not even sure if they're still used.

That said, these constants are sort of esoteric. I'm not sure we want to expose them to users, so I'd put them in a header that's only included from source files in that graphics system. I don't know that such a header exists, so you may need to add one.

**Comment by faissaloo** [comment]
We'd have to expose them to users in some way or having the macros in the first place is nigh-pointless because it would be too obscure for alot of users to bother with.

**Comment by JoshDreamland** [comment]
I'm missing some context, here. Let me just leave some comments inline.

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.cpp [hunk 1]
```diff
@@ -18,13 +18,17 @@
...
+#include "Graphics_Systems/General/GScolors.h"
 #include "Graphics_Systems/General/GSprimitives.h"
 #include "Graphics_Systems/General/GSbackground.h"
 #include "Graphics_Systems/General/GStextures.h"
 #include "Universal_System/nlpo2.h"
 #include "Universal_System/backgroundstruct.h"
 #include "Universal_System/spritestruct.h"
 
+//Note that this clamps between 0 and 1, not 0 and 255
+#define bind_alpha(alpha) (alpha <= 0 ? 0: alpha >= 1? 1: alpha)
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.cpp]: Nit: In graphics programming, `bind` has a very special meaning; I'd use `clamp`.

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.h [hunk 2]
```diff
@@ -19,22 +19,23 @@
...
 
 double background_get_texture_width_factor(int backId);
 double background_get_texture_height_factor(int backId);
-void draw_background(int back, gs_scalar x, gs_scalar y, int color = 16777215, gs_scalar alpha = 1);
-void draw_background_stretched(int back, gs_scalar x, gs_scalar y, gs_scalar wid, gs_scalar hei, int color = 16777215, gs_scalar alpha = 1);
-void draw_background_part(int back, gs_scalar left, gs_scalar top, gs_scalar wid, gs_scalar hei, gs_scalar x, gs_scalar y, int color = 16777215, gs_scalar alpha = 1);
-void draw_background_tiled(int back, gs_scalar x, gs_scalar y, int color = 16777215, gs_scalar alpha = 1);
-void draw_background_tiled_area(int back,gs_scalar x, gs_scalar y,gs_scalar x1, gs_scalar y1,float x2,float y2, int color = 16777215, gs_scalar alpha = 1);
-void draw_background_ext(int back, gs_scalar x, gs_scalar y, gs_scalar xscale = 1, gs_scalar yscale = 1, double rot = 0, int color = 16777215, gs_scalar alpha = 1);
+void draw_background(int back, gs_scalar x, gs_scalar y, int color = c_white, gs_scalar alpha = draw_get_alpha());
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.h]: This behavior is... disagreeable, but so is the GM6 behavior. Why are we defaulting the color but not the alpha? This is a confusing concept for people who haven't drank the Yoyo kool-aid. For that reason, I prefer the GM6 behavior. Either way, do we have compatibility macros? It would be a good idea to `#define DEFAULT_IMAGE_ALPHA` to either of `draw_get_alpha()` or `1` atop this file, conditioned on those macros. `#undef` it at the bottom, too.

Comment (faissaloo) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.h]: >Why are we defaulting the color but not the alpha?  
  
Because that's how it is in GMS, I don't think it's too unreasonable, the color is more accurately defined as 'blend color' rather than 'draw color' when you're dealing with images.  
  
>Either way, do we have compatibility macros? It would be a good idea to #define DEFAULT_IMAGE_ALPHA to either of draw_get_alpha() or 1 atop this file, conditioned on those macros. #undef it at the bottom, too.  
  
Aside from that, how do you do compatibility macros? Are they just #ifndef #define #endif definitions that get overridden at compile if needed?

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.cpp (hunk 1 above)
Comment (faissaloo) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.cpp]: On it.

Comment (faissaloo) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.cpp]: Done.

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.h (hunk 2 above)
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSbackground.h]: > I don't think it's too unreasonable, the color is more accurately defined as 'blend color' rather than 'draw color' when you're dealing with images.

In graphics programming, you either input a raw color to a vertex, or you don't. When alpha blending is in play, it's usually grouped with color blending. I think users will be weirded out to discover that we make some distinction, though maybe that seems natural in the Game Maker workflow.

> Are they just #ifndef #define #endif definitions that get overridden at compile if needed?

That's correct. I know sorlok added some GM5 compatibility macros, but I don't know if there are any for GM6/7/8.

**Comment by JoshDreamland** [comment]
Yeah, looking this over, I see what Robert was talking about. I would not expose those macros to users; I'd just configure them based on compatibility constants. My objection to exposing them to users is that they are not constants; they're macros that may expand to a function call. Moreover, they're a product of the configuration, which the user should be aware of.

So my suggested pattern for them is
```C++
#if GM_COMPATIBILITY_VERSION < 8100
#  define DEFAULT_ALPHA draw_get_alpha()
#else
#  define DEAFAULT_ALPHA 1
#endif

void draw_whatever(gs_scalar x, gs_scalar y /* ... */, float alpha = DEFAULT_ALPHA);
// ... More such functions ...

#undef DEFAULT_ALPHA
```

**Comment by faissaloo** [comment]
>My objection to exposing them to users is that they are not constants; they're macros that may expand to a function call. Moreover, they're a product of the configuration, which the user should be aware of.  
Right, got you.  
  
>Note that I completely made up that compatibility constant. One may or may not exist.  
  
So should I add a [COMPLIANCE_LVL](https://github.com/enigma-dev/enigma-dev/blob/03b754e1907fb840e4b745252333e504151bae7c/CompilerSource/settings.h#L58  ) for Game Maker 8? COMPL_GM8 or something? Also, does anyone have a copy of GM7 to test if the GM8 and GM7 behaviour is the same, so I know if I can just do a `|| compliance_mode==COMPL_GM7`. If we add another compliance level I don't think it would be a good idea to have a batch one for both GM7 and GM8 because we might discover some other subtle difference between the two later on.

**Comment by faissaloo** [comment]
I've been mulling over this for a bit and I think it's best that we make the compliance modes correspond to integers so we can do comparisons with them, similar to how @JoshDreamland suggested in his example, right now compliance modes are just assigned to numbers in the order that they were added, instead they should go like this so that they can easily be further expanded should we find anymore incompatibilities:  
```
enum COMPLIANCE_LVL {
    COMPL_STANDARD = 0,
    COMPL_GM5 = 5,
    COMPL_GM8 = 8
};
```  

Looking at the changelog for GM81 I don't see any incompatibilities, so I don't think there will be any need for a GM81 COMPLIANCE_LVL so I don't think there will be a need to take subversions into account. This shouldn't break anything because I don't see any code relying on COMPL_GM5 having a specific value.

**Comment by faissaloo** [comment]
How's this? (sorry some of the trailing tabs being removed, my editor did that, didn't realise until I pushed)

**Comment by JoshDreamland** [comment]
Looks fine; we'll sort out the values for those constants, later.

**Comment by faissaloo** [comment]
Either I forgot to commit or it got lost in the force push but this is missing so the compliance mode thing is now always going to use GM8 behaviour (because it compares with `<=8` ):  
```
enum COMPLIANCE_LVL {
    COMPL_STANDARD = ~0,
    COMPL_GM5 = 5, //now COMPL_GM567
    COMPL_GM8 = 8
};
```
Right now it's:  
```
  enum COMPLIANCE_LVL {
    COMPL_STANDARD = 0,    //Standard (enigma) compliance. Default and recommended.
    COMPL_GM567 = 1,       //GM5,6,7 compliance. timeline_running will default to "true". exit will abort single code actions.
    COMPL_GM8 = 2,         //GM8 compliance. exit will abort single code actions.
};
```

**Comment by JoshDreamland** [comment]
> There isn’t anything to compare.
> master is up to date with all commits from 11d74842e282890f5a2ee8dde2760d0a368785b3.
Seems everything's accounted for, so anything missing would have to have been explicitly reverted. Feel free to send the other commits in a pull request, and we'll get them in.
